sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -597,31 +623,11 @@ object LimitPushDown extends Rule[LogicalPlan] {
     // pushdown Limit.
     case LocalLimit(exp, u: Union) =>
       LocalLimit(exp, u.copy(children = u.children.map(maybePushLocalLimit(exp, _))))
-    // Add extra limits below JOIN:
-    // 1. For LEFT OUTER and RIGHT OUTER JOIN, we push limits to the left and right sides,
-    //    respectively.
-    // 2. For INNER and CROSS JOIN, we push limits to both the left and right sides if join
-    //    condition is empty.
-    // 3. For LEFT SEMI and LEFT ANTI JOIN, we push limits to the left side if join condition
-    //    is empty.
-    // It's not safe to push limits below FULL OUTER JOIN in the general case without a more
-    // invasive rewrite. We also need to ensure that this limit pushdown rule will not eventually
-    // introduce limits on both sides if it is applied multiple times. Therefore:
-    //   - If one side is already limited, stack another limit on top if the new limit is smaller.
-    //     The redundant limit will be collapsed by the CombineLimits rule.
-    case LocalLimit(exp, join @ Join(left, right, joinType, conditionOpt, _)) =>
-      val newJoin = joinType match {
-        case RightOuter => join.copy(right = maybePushLocalLimit(exp, right))
-        case LeftOuter => join.copy(left = maybePushLocalLimit(exp, left))
-        case _: InnerLike if conditionOpt.isEmpty =>
-          join.copy(
-            left = maybePushLocalLimit(exp, left),
-            right = maybePushLocalLimit(exp, right))
-        case LeftSemi | LeftAnti if conditionOpt.isEmpty =>
-          join.copy(left = maybePushLocalLimit(exp, left))
-        case _ => join
-      }
-      LocalLimit(exp, newJoin)
+
+    case LocalLimit(exp, join: Join) =>
+      LocalLimit(exp, pushLocalLimitThroughJoin(exp, join))
+    case LocalLimit(exp, project @ Project(_, join: Join)) =>
```

Comment:
Thanks @wangyum for adding this. I think there might be a list of operators safe to push through besides `Project` - e.g. `Sort`, `RepartitionByExpression`, `ScriptTransformation`, etc.

Shall we add push down through `Project` separately? It should not be only restricted to `Project(Join)` right? It can be `Project(OtherOperator)` as well?

```
case LocalLimit(exp, p: Project) =>
  LocalLimit(exp, p.copy(maybePushLocalLimit(exp, _)))
```

Reply:
I change it to `case LocalLimit(exp, project @ Project(_, child)) if !child.isInstanceOf[LeafNode] =>` to avoid do some useless work if direct pushdown limit through Project:
```
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.LimitPushDown ===
 GlobalLimit 5                                                GlobalLimit 5
 +- LocalLimit 5                                              +- LocalLimit 5
    +- Project [a#0]                                             +- Project [a#0]
!      +- Join LeftOuter, ((a#0 = x#3) AND (b#1 = y#4))             +- LocalLimit 5
!         :- LocalLimit 5                                              +- Join LeftOuter, ((a#0 = x#3) AND (b#1 = y#4))
!         :  +- Project [a#0, b#1]                                        :- LocalLimit 5
!         :     +- Relation default.t1[a#0,b#1,c#2] parquet               :  +- Project [a#0, b#1]
!         +- Project [x#3, y#4]                                           :     +- LocalLimit 5
!            +- Filter (isnotnull(x#3) AND isnotnull(y#4))                :        +- Relation default.t1[a#0,b#1,c#2] parquet
!               +- Relation default.t2[x#3,y#4,z#5] parquet               +- Project [x#3, y#4]
!                                                                            +- Filter (isnotnull(x#3) AND isnotnull(y#4))
!                                                                               +- Relation default.t2[x#3,y#4,z#5] parquet


=== Applying Rule org.apache.spark.sql.catalyst.optimizer.CollapseProject ===
 GlobalLimit 5                                                            GlobalLimit 5
 +- LocalLimit 5                                                          +- LocalLimit 5
!   +- Project [a#0]                                                         +- LocalLimit 5
!      +- LocalLimit 5                                                          +- Project [a#0]
!         +- Project [a#0]                                                         +- Join LeftOuter, ((a#0 = x#3) AND (b#1 = y#4))
!            +- Join LeftOuter, ((a#0 = x#3) AND (b#1 = y#4))                         :- LocalLimit 5
!               :- LocalLimit 5                                                       :  +- LocalLimit 5
!               :  +- Project [a#0, b#1]                                              :     +- Project [a#0, b#1]
!               :     +- LocalLimit 5                                                 :        +- Relation default.t1[a#0,b#1,c#2] parquet
!               :        +- Project [a#0, b#1]                                        +- Project [x#3, y#4]
!               :           +- Relation default.t1[a#0,b#1,c#2] parquet                  +- Filter (isnotnull(x#3) AND isnotnull(y#4))
!               +- Project [x#3, y#4]                                                       +- Relation default.t2[x#3,y#4,z#5] parquet
!                  +- Filter (isnotnull(x#3) AND isnotnull(y#4))          
!                     +- Relation default.t2[x#3,y#4,z#5] parquet         
           
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.EliminateLimits ===
 GlobalLimit 5                                                      GlobalLimit 5
!+- LocalLimit 5                                                    +- LocalLimit least(5, 5)
!   +- LocalLimit 5                                                    +- Project [a#0]
!      +- Project [a#0]                                                   +- Join LeftOuter, ((a#0 = x#3) AND (b#1 = y#4))
!         +- Join LeftOuter, ((a#0 = x#3) AND (b#1 = y#4))                   :- LocalLimit least(5, 5)
!            :- LocalLimit 5                                                 :  +- Project [a#0, b#1]
!            :  +- LocalLimit 5                                              :     +- Relation default.t1[a#0,b#1,c#2] parquet
!            :     +- Project [a#0, b#1]                                     +- Project [x#3, y#4]
!            :        +- Relation default.t1[a#0,b#1,c#2] parquet               +- Filter (isnotnull(x#3) AND isnotnull(y#4))
!            +- Project [x#3, y#4]                                                 +- Relation default.t2[x#3,y#4,z#5] parquet
!               +- Filter (isnotnull(x#3) AND isnotnull(y#4))       
!                  +- Relation default.t2[x#3,y#4,z#5] parquet      
 
```

Reply:
Sounds like a good idea. We can have an allowlist of unary operators.

Reply:
This will introduce useless pushdown. For example:
```scala
spark.range(200L).selectExpr("id AS a", "id AS b").write.saveAsTable("t1")
spark.range(300L).selectExpr("id AS x", "id AS y").write.saveAsTable("t2")
spark.sql("SELECT 1 FROM t1 INNER JOIN t2 ON a = x limit 10").explain(true)
```

```
== Optimized Logical Plan ==
GlobalLimit 10
+- LocalLimit 10
   +- Project [1 AS 1#20]
      +- LocalLimit 10
         +- Project
            +- Join Inner, (a#16L = x#18L)
               :- Project [a#16L]
               :  +- Filter isnotnull(a#16L)
               :     +- Relation default.t1[a#16L,b#17L] parquet
               +- Project [x#18L]
                  +- Filter isnotnull(x#18L)
                     +- Relation default.t2[x#18L,y#19L] parquet

== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- CollectLimit 10
   +- Project [1 AS 1#20]
      +- LocalLimit 10
         +- Project
            +- BroadcastHashJoin [a#16L], [x#18L], Inner, BuildLeft, false
               :- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, false]),false), [id=#68]
               :  +- Filter isnotnull(a#16L)
               :     +- FileScan parquet default.t1[a#16L] 
               +- Filter isnotnull(x#18L)
                  +- FileScan parquet default.t2[x#18L]

```

Another example is TPC-DS q32:
https://github.com/apache/spark/blob/66f5a42ca5d259038f0749ae2b9a04cc2f658880/sql/core/src/test/resources/tpcds/q32.sql#L1-L15

Reply:
@c21 @maropu @cloud-fan 
I think `Project` is the most used, and this will not introduce useless `LocalLimit`.